### PR TITLE
Tensorflow batchnorm float16 support

### PR DESCRIPTION
**Pull request by bradklingensmith**
This adds support for properly using the Tensorflow fused_batchnorm API (passing in float32 values for scale and offset). And when Tensorflow is the backend, it follows NVIDIA's recommendation of storing weights in float32, performing the weight updates in float32, and performing aggregations in float32.

### CODE REVIEW

**Comment by lexxxel** [comment]
I rebased his commits. It would be nice to see this MR merged soon :-)

**Comment by fchollet** [comment]
Yes, sorry I haven't had time to review it yet. Can you find an experience Keras user/developer to review it?

Also, there seems to be some merge conflicts.

**Comment by fchollet** [comment]
Closing outdated PR. Sorry we couldn't find available reviewers to process it in time.

By the way, this PR seems to be doing mixed-precision for BN (only), but doing mixed precision correctly would involve pretty extensive changes to the API in general. We will support it in tf.keras in the future; no plans for multi-backend Keras yet.

**Comment by bradklingensmith** [comment]
To clarify, it only does mixed-precision for BN because BN doesn't function well during training when the aggregation variables (mean and standard deviation) are in FP16. So, per NVIDIA's suggestion, when you want to train an FP16 network, you should leave the aggregation variables in FP32. And Tensorflow does exactly that. So Keras can't currently train an FP16 network with a BN layer and Tensorflow as the backend. That's why the pull request is named Tensorflow batchnorm float16 support and not Tensorflow batchnorm mixed-precision support. And this isn't Tensorflow specific. Supporting FP16 BN on any backend is going to require this.
https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html
